cpp/ql/src/experimental/Security/CWE/CWE-416/IteratorToExpiredContainer.ql

```diff
@@ -86,12 +70,39 @@ predicate destroyedToBeginSink(DataFlow::Node sink, FunctionCall fc) {
 }
 
 /**
- * Flow from any destroyed object to the qualifier of a `begin` or `end` call
+ * A configuration to track flow from a temporary variable to the qualifier of
+ * a destructor call, and subsequently to a qualifier of a call to `begin` or
+ * `end`.
  */
-module DestroyedToBeginConfig implements DataFlow::ConfigSig {
-  predicate isSource(DataFlow::Node source) { source = getADestroyedNode() }
+module Config implements DataFlow::StateConfigSig {
+  newtype FlowState =
+    additional TempToDestructor() or
+    additional DestroyedToBegin(DataFlow::Node n) {
+      exists(DataFlow::Node thisOperand |
+        tempToDestructorSink(thisOperand, _) and
+        n = getADestroyedNode(thisOperand)
+      )
+    }
+
+  predicate isSource(DataFlow::Node source, FlowState state) {
+    source.asInstruction().(VariableAddressInstruction).getIRVariable() instanceof IRTempVariable and
+    state = TempToDestructor()
+  }
+
+  predicate isAdditionalFlowStep(
+    DataFlow::Node node1, FlowState state1, DataFlow::Node node2, FlowState state2
+  ) {
+    tempToDestructorSink(node1, _) and
+    state1 = TempToDestructor() and
+    state2 = DestroyedToBegin(node2) and
+    node2 = getADestroyedNode(node1)
+  }
 
-  predicate isSink(DataFlow::Node sink) { destroyedToBeginSink(sink, _) }
+  predicate isSink(DataFlow::Node sink, FlowState state) {
+    // Note: This is a non-trivial cartesian product!
+    // Hopefully, both of these sets are quite small in practice
```

Comment:
I was a tiny bit concerned that in a database with lots of results for this query, both sets might be distinctly "medium" rather than clearly "small".  Mind you a cartesian product of even a few thousand nodes / states still shouldn't be enough to matter.

Based on the stage timings table on the DCA run, I decided to have a closer look at an `abseil-cpp-linux` run locally.  It took one second to execute (with a cache warmed up on another dataflow query), no more than 122ms on any single predicate.  So I'm reassured.

Reply:
Indeed, this is basically the same CP as the one we fixed for a different query in https://github.com/github/codeql/pull/10123. We could make this a lot safer by doing a similar transformation, but I didn't want to do it before we had seen a DB that needed this.

Reply:
We can also consider doing this transformation before we pull the query out of experimental. But in any case, I would prefer to keep that transformation out of this PR

Reply:
Sounds like it might be worth doing before we pull the query out of experimental, or alternatively we can rely on QA to catch any problems.  Either way you're right, it doesn't need addressing here.